yarn-project/simulator/src/avm/avm_simulator.ts

```diff
@@ -110,8 +136,45 @@ export class AvmSimulator {
       // Note: "exceptional halts" cannot return data, hence []
       const results = new AvmContractCallResult(/*reverted=*/ true, /*output=*/ [], revertReason);
       this.log.debug(`Context execution results: ${results.toString()}`);
+
+      this.printOpcodeTallies();
       // Return results for processing by calling context
       return results;
     }
   }
+
+  private tallyInstruction(pc: number, opcode: string, gasUsed: Gas) {
+    const opcodeTally = this.opcodeTallies.get(opcode) || ({ count: 0, gas: { l2Gas: 0, daGas: 0 } } as OpcodeTally);
+    opcodeTally.count++;
+    opcodeTally.gas.l2Gas += gasUsed.l2Gas;
+    opcodeTally.gas.daGas += gasUsed.daGas;
+    this.opcodeTallies.set(opcode, opcodeTally);
+
+    const pcTally = this.pcTallies.get(pc) || ({ opcode: opcode, count: 0, gas: { l2Gas: 0, daGas: 0 } } as PcTally);
+    pcTally.count++;
+    pcTally.gas.l2Gas += gasUsed.l2Gas;
+    pcTally.gas.daGas += gasUsed.daGas;
+    this.pcTallies.set(pc, pcTally);
+  }
+
+  private printOpcodeTallies() {
+    this.log.debug(`Printing tallies per opcode sorted by gas...`);
+    // sort descending by L2 gas consumed
+    const sortedOpcodes = Array.from(this.opcodeTallies.entries()).sort((a, b) => b[1].gas.l2Gas - a[1].gas.l2Gas);
+    for (const [opcode, tally] of sortedOpcodes) {
+      // NOTE: don't care to clutter the logs with DA gas for now
+      this.log.debug(`${opcode} executed ${tally.count} times consuming a total of ${tally.gas.l2Gas} L2 gas`);
+    }
+
+    this.log.debug(`Printing tallies per PC sorted by #times each PC was executed...`);
+    const sortedPcs = Array.from(this.pcTallies.entries())
+      .sort((a, b) => b[1].count - a[1].count)
+      .filter((_, i) => i < 20);
+    for (const [pc, tally] of sortedPcs) {
+      // NOTE: don't care to clutter the logs with DA gas for now
+      this.log.debug(
+        `PC:${pc} containing opcode ${tally.opcode} executed ${tally.count} times consuming a total of ${tally.gas.l2Gas} L2 gas`,
```

Comment:
Does `${tally.opcode}` print the number or the name? I recall having used/seen something like `${OpCode[tally.opcode]}` to get the name, but I don't remember the language XD

Reply:
Prints the name because it's actually just constructor.name. When I used instruction.opcode, i was getting undefined for a bunch of instructions